### PR TITLE
Add differentiable functions doc.

### PR DESCRIPTION
**Pull request by dan-zheng**
Add "Differentiable functions and differentiation APIs" document, explaining:
- `@differentiable` function types.
- `@differentiable` declarations.
- differentiation APIs.

### CODE REVIEW

File: docs/DifferentiableFunctions.md [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+let x: Float = 3.0
+
+// Free function examples.
+
+// Computes the gradient of `square` at `x`.
+print(gradient(at: x, in: square)) // 6.0
+// Computes the gradient of `square`, then applies it to `x`.
+print(gradient(of: square)(x)) // 6.0
+// Computes the value and gradient of `square` at `x`.
+print(valueWithGradient(at: x, in: square)) // (value: 9.0, gradient: 6.0)
```
Comment (dan-zheng) [inline docs/DifferentiableFunctions.md]: It may be nice to include examples of `pullback` and `valueWithPullback`.
Also, it may be nice to include examples of differentiation APIs for higher arity functions.

File: docs/DifferentiableFunctions.md [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+## `@differentiable` function type attribute
+
+### Basics
+
+In Swift, function types can have attributes. When a function type is annotated
+with `@differentiable`, Swift guarantees that all values of that function type
+can be differentiated.
+
+`@differentiable` functions can be called like normal functions, or be passed to
+APIs that take `@differentiable` functions, like [`gradient(of:)`]. The binary
```
Comment (bartchr808) [inline docs/DifferentiableFunctions.md]: The `gradient(of:)` seems to be linking to this 
https://tensorflow.devsite.corp.google.com/swift/api_docs/Functions#/s:10TensorFlow8gradient2of15CotangentVectorQzxcAA0A0Vyq_Gxc_tAA14DifferentiableRzAA0aB13FloatingPointR_r0_lF 
which seems inaccessible

Comment (bartchr808) [reply docs/DifferentiableFunctions.md]: This is true for the links to the differentiation APIs table as well

Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Thanks for the catch! Fixed.

File: docs/DifferentiableFunctions.md [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+```
+
+There are a few reasons why differentiation can fail:
+* The function to differentiate contains non-differentiable computation along
+  the dataflow from parameters to function result.
+* The function to differentiate is defined in another module.
+* The function to differentiate uses [control
+  flow](https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html)
+  (if-statements, switch-statements, loops, etc). This restriction will be
+  lifted soon.
```
Comment (bartchr808) [inline docs/DifferentiableFunctions.md]: Can't wait to help out with this! 😄

File: docs/DifferentiableFunctions.md [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+```
+
+All other differentiation APIs are defined in terms of `valueWithPullback`.
+Here's an example for `gradient`:
+
+```swift
+// `gradient` returns the partial derivative with respect to varying parameters for scalar-result
+// functions. It simply returns `pullback(1)`.
+func fooGradient(_ x: Float) -> Float {
+    let pullback = foo_valueWithPullback(x).pullback
```
Comment (bartchr808) [inline docs/DifferentiableFunctions.md]:
```suggestion
    let pullback = fooValueWithPullback(x).pullback
```

**Comment by rxwei** [comment]
Great progress, I’ll have some time to review this later today!

File: docs/DifferentiableFunctions.md (hunk 4 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Thanks! Fixed in https://github.com/tensorflow/swift/pull/145/commits/a5cd494b0ec9ad754b793eaebffd7417987820a4.

File: docs/DifferentiableFunctions.md [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+## Constructing a `@differentiable` function from a derivative
+
+Given a function and its derivative, it is possible to construct a
+`@differentiable` version of the function using the
+`differentiableFunction(from:)` helper function defined in the standard library.
+
+Here's an example:
+```swift
+let multiply: @differentiable (Float, Float) -> Float =
+    differentiableFunction(from: { x, y in (value: x * y, pullback: { v in v * y, v * x }))
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Syntax error on the pullback closure. Don’t think this will compile.

File: docs/DifferentiableFunctions.md [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+print(gradient(of: sillyExp)(3))
+// Taking 𝑒(2.7182817) to the power of 3.0!
+// 20.085535
+```
+
+## Constructing a `@differentiable` function from a derivative
+
+Given a function and its derivative, it is possible to construct a
+`@differentiable` version of the function using the
+`differentiableFunction(from:)` helper function defined in the standard library.
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Link to API docs.

File: docs/DifferentiableFunctions.md [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+test.swift:1:52: error: function is not differentiable
+let add: @differentiable (Float, Float) -> Float = { x, y in
+                                                   ^~~~~~~~~
+test.swift:2:11: note: cannot differentiate through a non-differentiable result; do you want to add '.withoutDerivative()'?
+    Float(Int(x + y))
+          ^
+```
+
+There are a few reasons why differentiation can fail:
+* The function to differentiate contains non-differentiable computation along
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Earlier Conal Elliott on Twitter made a good point about the distinction between “non-differentiable functions” (mathematically non differentiable functions) and “functions whose derivatives are not computable” (opaque functions). Since our language doesn’t know math, we don’t really know what functions are really nondifferentiable in the mathematical sense. So we’ve been using “cannot differentiate xxx” in our diagnostics instead of “xxx is nondifferentiable”. We should probably use the same wording here instead of “nondifferentiable”.

File: docs/DifferentiableFunctions.md (hunk 5 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Thanks for the catch! Fixed.

File: docs/DifferentiableFunctions.md (hunk 6 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Done.

File: docs/DifferentiableFunctions.md (hunk 7 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: How about:
`* The function to differentiate contains dataflow, from parameters to result, that cannot be differentiated.`

Comment (rxwei) [reply docs/DifferentiableFunctions.md]: “Data flow” or “dataflow” is a compiler term IMO. The user understands “code” or “computation”.

Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: How about:
```
* The function to differentiate contains computation, from parameters to result, that cannot be differentiated.
```

And also, how about adding another point for opaque functions?
```
* The function to differentiate is opaque, i.e. it is a function parameter with a non-`@differentiable` function type.
```

Comment (rxwei) [reply docs/DifferentiableFunctions.md]: I think both these are good. Any inability to differentiate something always comes from encountering an opaque, non-@differentiable function.

File: docs/DifferentiableFunctions.md [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+func sillyExp(_ x: Float) -> Float {
+    let 𝑒 = Float(M_E)
+    print("Taking 𝑒(\(𝑒)) to the power of \(x)!")
+    return pow(𝑒, x)
+}
+
+@differentiating(sillyExp)
+func sillyDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+    let y = sillyExp(x)
+    return (value: y, pullback: { _ in y })
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Pullback should be v * y.

Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Fixed, thanks.

File: docs/DifferentiableFunctions.md (hunk 7 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Added both points.

File: docs/DifferentiableFunctions.md [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+}
+```
+
+All other differentiation APIs are defined in terms of `valueWithPullback`.
+Here's an example for `gradient`:
+
+```swift
+// `gradient` returns the partial derivative with respect to varying parameters for scalar-result
+// functions. It simply returns `pullback(1)`.
+func fooGradient(_ x: Float) -> Float {
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: The fact that 1) this is named something other than `gradient` and 2) this is not a differentiation API is a bit misleading. I'd suggest showing a completely generic `gradient(of:)` implementation here, and provide a link to the standard library implementation.

File: docs/DifferentiableFunctions.md [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+```swift
+extension Float {
+    @differentiable
+    public static func + (lhs: Float, rhs: Float) -> Float { ... }
+}
+```
+
+Besides function declarations, there are a few other function-like declarations
+that can be marked with `@differentiable`:
+- Computed property getters. (This requires both the type defining the property
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: It's not just computed properties--same for stored properties, though their getter is not user-definable. You don't have to mention "getter" except to mention that this is making the getter operation differentiable w.r.t. `self`.

Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Good point. I changed it to the following (didn't mention setters):
```
- Stored and computed properties.
  - This requires both the type defining the property
    and the type of the property to conform to `Differentiable`.
  - Property getters are differentiable with respect to `self`.
```

File: docs/DifferentiableFunctions.md (hunk 9 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: I see - showing a generic `gradient(of:)` implementation using `valueWithPullback` is indeed more insightful.

Done.

File: docs/DifferentiableFunctions.md [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+//   - bias: [1.0, 1.0]
+print(𝛁input)
+// [[2.0, 2.0]]
+```
+
+## Providing a custom derivative
+
+Use the `@differentiating` attribute to mark a function as a custom derivative
+for another function. This is useful for registering derivatives for primitive
+operations.
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: "primitive operations" -> "functions that cannot be differentiated by traversing the function body, typically a primitive math operator in a library"

File: docs/DifferentiableFunctions.md [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+    }
+    return original
+}
+```
+
+## `@differentiable` functions and automatic differentiation
+
+Automatic differentiation is the technique used by the compiler to automatically
+compute function derivatives. This document does not go into detail about
+automatic differentiation - but with an understanding of `@differentiable`
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]:
```suggestion
the automatic differentiation algorithm - but with an understanding of `@differentiable`
```

File: docs/DifferentiableFunctions.md [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+    func derivative(_ x: T) -> (value: R, pullback: (R.CotangentVector) -> T.CotangentVector) {
+        return vjp(x)
+    }
+    return original
+}
+```
+
+## `@differentiable` functions and automatic differentiation
+
+Automatic differentiation is the technique used by the compiler to automatically
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Add link: https://en.wikipedia.org/wiki/Automatic_differentiation

File: docs/DifferentiableFunctions.md [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+compute function derivatives. This document does not go into detail about
+automatic differentiation - but with an understanding of `@differentiable`
+functions and differentiation APIs, one can get a glimpse of how automatic
+differentiation works.
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Perhaps add some lines to imply to our users that "automatic differentiation" is just an implementation detail in the compiler that enables derivative calculation, and that only differentiation APIs and `@differentiable` attributes are what's to be internalized.

File: docs/DifferentiableFunctions.md [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+
+The key differentiation API is the [`valueWithPullback`] function, which takes a
+`@differentiable` function and arguments and returns two things: the result of
+the function when applied to arguments, and a backpropagation function called a
+"pullback", which takes the gradient of the result and returns the gradient of
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Reword:
"... which takes a direction vector (usually the partial derivative w.r.t. the original result) and returns the directional gradient w.r.t. the varying parameters.

File: docs/DifferentiableFunctions.md [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+
+## `@differentiable` functions and automatic differentiation
+
+Automatic differentiation is the technique used by the compiler to automatically
+compute function derivatives. This document does not go into detail about
+automatic differentiation - but with an understanding of `@differentiable`
+functions and differentiation APIs, one can get a glimpse of how automatic
+differentiation works.
+
+The key differentiation API is the [`valueWithPullback`] function, which takes a
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]:
```suggestion
The key differentiation API is the [`valueWithPullback(at:in:)`] function, which takes a
```

File: docs/DifferentiableFunctions.md [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+
+```swift
+func foo(_ x: Float) -> Float {
+    let double = x + x
+    let result = double * double
+    return result
+}
+```
+
+Conceptually, here's how the compiler computes `valueWithPullback` for `foo`:
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]:
```suggestion
Conceptually, here's how the compiler computes the `value` and the `pullback` for `foo`:
```

File: docs/DifferentiableFunctions.md [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+All other differentiation APIs are defined in terms of `valueWithPullback`.
+For example, here's how [`gradient(at:in:)`] is defined:
+
+```swift
+// `gradient` returns the partial derivative with respect to varying parameters for scalar-result
+// functions. It simply returns `pullback(1)`.
+func gradient<T, R>(
+    at x: T, in f: @differentiable (T) -> R
+) -> T.CotangentVector
+    where T : Differentiable, R : FloatingPoint & Differentiable,
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]:
```suggestion
    where T: Differentiable, R: FloatingPoint & Differentiable,
```

File: docs/DifferentiableFunctions.md [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+          R.CotangentVector == R {
+    let (value, pullback) = valueWithPullback(at: x, in: f)
+    return pullback(R(1))
+}
+
+print(gradient(at: x, in: foo)) // 24.0
+```
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: Add a line below or above: 
"You can find [its implementation](https://github.com/apple/swift/blob/6e20cf1d2746435d879ccfeee9713aa7a32e1081/stdlib/public/core/AutoDiff.swift#L467) in the standard library."

File: docs/DifferentiableFunctions.md [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+### Basics
+
+The `@differentiable` attribute can also be applied to function declarations.
+`@differentiable` marks a function as being differentiable with respect to some
+parameters (the varying parameters, explained below). `@differentiable` requires
+the types of the varying parameters and the function result type to all conform
+to the `Differentiable` protocol.
+
+This annotation does not change the declaration to have a `@differentiable`
+function type; instead, it triggers differentiation by the compiler on the
```
[the remark this answers is not in the document]
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Changing to:
```
This annotation does not change the function declaration to have a
`@differentiable` function type; instead, it tells the compiler to differentiate
the function when compiling its containing file.
```

File: docs/DifferentiableFunctions.md (hunk 11 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Changing to:
```
Use the `@differentiating` attribute to mark a function as a custom derivative
for another function. This is useful for registering derivatives for functions
that cannot be differentiated by traversing the function body, typically
primitive math library operators.
```

File: docs/DifferentiableFunctions.md [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+
+```swift
+@differentiable(wrt: (input, filter))
+func conv2d(input: Tensor<Float>, filter: Tensor<Float>, strides: (Int, Int), padding: Padding) {
+    ...
+}
+```
+
+Functions can have multiple `@differentiable` attributes with differentiable
+`wrt` parameter lists. `@differentiable` protocol requirements If a protocol
```
[the remark this answers is not in the document]
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Oops, thanks!

File: docs/DifferentiableFunctions.md (hunk 19 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Adding:
```markdown
You can find [the implementation of
`gradient(at:in:)`](https://github.com/apple/swift/blob/6e20cf1d2746435d879ccfeee9713aa7a32e1081/stdlib/public/core/AutoDiff.swift#L467)
in the standard library.
```

File: docs/DifferentiableFunctions.md (hunk 15 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: That's much better, thanks!

File: docs/DifferentiableFunctions.md (hunk 14 above)
Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Adding:
```markdown
*Note: automatic differentiation is an implementation detail in the compiler
that enables differentiation. `@differentiable` functions and differentiation
APIs are the important, visible features to understand.*
```

**Comment by dan-zheng** [comment]
Thanks @rxwei for the thorough review!
All comments have addressed, ready for re-review.

File: docs/DifferentiableFunctions.md [hunk 22]
```diff
@@ -0,0 +1,468 @@
...
+```
+
+## Providing a custom derivative
+
+Use the `@differentiating` attribute to mark a function as a custom derivative
+for another function. This is useful for registering derivatives for functions
+that cannot be differentiated by traversing the function body, typically
+primitive math library operators.
+
+Note: currently, the `@differentiating` attribute can only be used to define
```
Comment (rxwei) [inline docs/DifferentiableFunctions.md]: The other note paragraph is italic. Do you want to unify the style?

Comment (dan-zheng) [reply docs/DifferentiableFunctions.md]: Done: italicized both notes.

**Comment by rxwei** [comment]
I think the document is good to go!
